python/GafferUI/CompoundEditor.py

```diff
@@ -543,37 +562,89 @@ def __updatePinningButton( self, *unused ) :
 
 			self.__pinningButton.setVisible( True )
 
-			if editor.getNodeSet().isSame( editor.scriptNode().selection() ) :
-				self.__pinningButton.setToolTip( "Click to lock view to current selection" )
-				self.__pinningButton.setImage( "targetNodesUnlocked.png" )
+			drivingEditor, mode = editor.getNodeSetDriver()
+
+			if drivingEditor is not None :
+				icon = "nodeSetDriver%s.png" % mode
+				tip = "Click to un-link and follow the current node selection"
+			elif self.__editorIsPinned( editor ):
+				nodeSet = editor.getNodeSet()
+				icon = "nodeSet%s.png"  % nodeSet.__class__.__name__
+				tip = "Click to un-pin and follow the current node selection"
 			else :
-				self.__pinningButton.setToolTip( "Click to unlock view and follow selection" )
-				self.__pinningButton.setImage( "targetNodesLocked.png" )
+				icon = "nodeSetDriverGraphSelection.png"
+				tip = "Click to pin the current node selection"
+
+			self.__pinningButton.setToolTip( tip )
+			self.__pinningButton.setImage( icon )
 
 		else :
 
 			self.__pinningButton.setVisible( False )
 
+		self.__updateDrivenEditorSwatch( True )
+
+	def __updateDrivenEditorSwatch( self, updateAssociated, *args ) :
+
+		self.__drivenEditorSwatch.update()
+
+		if updateAssociated :
+
+			editor = self.getCurrent()
+			if editor and isinstance( editor, GafferUI.NodeSetEditor ) :
+				# Technically, their driven status hasn't changed, but the color for them
+				# may have, eg: A -> B -> C and we just unlinked B from C. A should now show
+				# B's color not C's. We can't rely on nodeSetChangedSignal as the actual
+				# node set may not have changed....
+				for drivenEditor in editor.drivenNodeSets( recurse = True ) :
+					parentTabbedContainer = drivenEditor.ancestor( _TabbedContainer )
+					if parentTabbedContainer is not None :
+						parentTabbedContainer.__updateDrivenEditorSwatch( False )
+
 	def __pinningButtonClicked( self, button ) :
 
 		editor = self.getCurrent()
 		assert( isinstance( editor, GafferUI.NodeSetEditor ) )
 
-		nodeSet = editor.getNodeSet()
-		selectionSet = editor.scriptNode().selection()
-		if nodeSet.isSame( selectionSet ) :
-			nodeSet = Gaffer.StandardSet( list( nodeSet ) )
+		drivingEditor, _ = editor.getNodeSetDriver()
+		if drivingEditor :
+			self.__followGraphEditorSelection()
+		elif self.__editorIsPinned( editor ) :
+			self.__followGraphEditorSelection()
 		else :
-			nodeSet = selectionSet
-		editor.setNodeSet( nodeSet )
+			self.__pinToGraphEditorSelection()
+
+	def __editorIsPinned( self, editor ) :
+
+		# Linking trumps pinning determination
+		drivingEditor, _ = editor.getNodeSetDriver()
+		if drivingEditor is not None :
+			return False
+
+		if editor.getNodeSet().isSame( editor.scriptNode().selection() ) :
+			return False
+
+		# All states other than having a driving editor, or literally the
+		# ScriptNode selection are considered 'pinned'
+		return True
+
+	def __pinToGraphEditorSelection( self, *args ) :
```

Comment:
GraphEditor -> Script? The GraphEditor doesn't own the selection, and we want the CompoundEditor to have as little knowledge of individual editors as possible.

Reply:
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/dacd021d21f96b8465056a36b6f92ff76dc1b658